pkg/svcat/service-catalog/instance_test.go

```diff
@@ -0,0 +1,487 @@
+				return true, nil, fmt.Errorf(errorMessage)
+			})
+			sdk.ServiceCatalogClient = badClient
+
+			instances, err := sdk.RetrieveInstancesByPlan(plan)
+			Expect(instances).To(BeNil())
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring(errorMessage))
+			actions := badClient.Actions()
+			Expect(actions[0].Matches("list", "serviceinstances")).To(BeTrue())
+			opts := fields.Set{"spec.clusterServicePlanRef.name": plan.Name}
+			Expect(actions[0].(testing.ListActionImpl).GetListRestrictions().Fields.Matches(opts)).To(BeTrue())
+		})
+	})
+	Describe("InstanceParentHierarchy", func() {
+		It("calls the v1beta1 generated Get function repeatedly to build the heirarchy of the passed in service isntance", func() {
+			broker := &v1beta1.ClusterServiceBroker{ObjectMeta: metav1.ObjectMeta{Name: "foobar_broker"}}
+			class := &v1beta1.ClusterServiceClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "foobar_class",
+				},
+				Spec: v1beta1.ClusterServiceClassSpec{
+					ClusterServiceBrokerName: broker.Name,
+				},
+			}
+			plan := &v1beta1.ClusterServicePlan{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "foobar_plan",
+				},
+				Spec: v1beta1.ClusterServicePlanSpec{},
+			}
+			si = &v1beta1.ServiceInstance{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foobar",
+					Namespace: "foobar_namespace",
+				},
+				Spec: v1beta1.ServiceInstanceSpec{
+					ClusterServicePlanRef: &v1beta1.ClusterObjectReference{
+						Name: plan.Name,
+					},
+					ClusterServiceClassRef: &v1beta1.ClusterObjectReference{
+						Name: class.Name,
+					},
+				},
+			}
+			linkedClient := fake.NewSimpleClientset(si, si2, class, plan, broker)
+			sdk.ServiceCatalogClient = linkedClient
+			retClass, retPlan, retBroker, err := sdk.InstanceParentHierarchy(si)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(retClass.Name).To(Equal(class.Name))
+			Expect(retPlan.Name).To(Equal(plan.Name))
+			Expect(retBroker.Name).To(Equal(broker.Name))
+			actions := linkedClient.Actions()
+			getClass := testing.GetActionImpl{
+				ActionImpl: testing.ActionImpl{
+					Verb: "get",
+					Resource: schema.GroupVersionResource{
+						Group:    "servicecatalog.k8s.io",
+						Version:  "v1beta1",
+						Resource: "clusterserviceclasses",
+					},
+				},
+				Name: class.Name,
+			}
+			getPlan := testing.GetActionImpl{
+				ActionImpl: testing.ActionImpl{
+					Verb: "get",
+					Resource: schema.GroupVersionResource{
+						Group:    "servicecatalog.k8s.io",
+						Version:  "v1beta1",
+						Resource: "clusterserviceplans",
+					},
+				},
+				Name: plan.Name,
+			}
+			getBroker := testing.GetActionImpl{
+				ActionImpl: testing.ActionImpl{
+					Verb: "get",
+					Resource: schema.GroupVersionResource{
+						Group:    "servicecatalog.k8s.io",
+						Version:  "v1beta1",
+						Resource: "clusterservicebrokers",
+					},
+				},
+				Name: broker.Name,
+			}
+			Expect(actions).Should(ContainElement(getClass))
```

Comment:
Is this what you were referring to by non-deterministic? We can't test the order of the actions, so we are using a contains instead? LGTM

Reply:
Yeah, I can't assert actions[0] matches 'get, classes' because they don't always happen in the same order due to the go routines.